### PR TITLE
Add new rule PERM_SUPER_NOT_CALLED_IN_GETPERMISSIONS

### PR DESCRIPTION
**Pull request by baloghadamsoftware**
SEI CERT rule SEC07-J requires that the superclass's getPermissions() is
called when writing a custom class loader. This method is defined in
class java.net.URLClassLoader. Thus any class inheriting directly or
indirectly form URLClassLoader and overriding its getPermissions()
method must call the superclass's getPermissions() method in this
method. It is not enoguh to just call it but the the return value (an
instance of java.security.PermissionCollection) must be initialized by
calling the superclass's method instead of creating a new instance from
scratch.

This patch adds a new detector to check this rule.

----

Make sure these boxes are checked before submitting your PR -- thank you!

- [x] Added an entry into `CHANGELOG.md` if you have changed SpotBugs code

### CODE REVIEW

**Comment by baloghadamsoftware** [comment]
Sorry for force pushing but there was a git mess which I had to sort out.

**Comment by baloghadamsoftware** [comment]
No false positives at all. But also no findings (as expected) on several open-source projects:

[MATSim](https://codechecker-demo.eastus.cloudapp.azure.com/java-static-analysis/reports?run=SpotBugs%20on%20MATSim-Libs%20%28user%3A%20spotbugs,%20branch%3A%20master%29&newcheck=SpotBugs%20on%20MATSim-Libs%20%28user%3A%20baloghadamsoftware,%20branch%3A%20PermissionsSuper%29)
[SpotBugs itself](https://codechecker-demo.eastus.cloudapp.azure.com/java-static-analysis/reports?run=SpotBugs%20on%20SpotBugs%20%28user%3A%20spotbugs,%20branch%3A%20master%29&newcheck=SpotBugs%20on%20SpotBugs%20%28user%3A%20baloghadamsoftware,%20branch%3A%20PermissionsSuper%29)
[Jenkins](https://codechecker-demo.eastus.cloudapp.azure.com/java-static-analysis/reports?run=SpotBugs%20on%20Jenkins%20%28user%3A%20spotbugs,%20branch%3A%20master%29&newcheck=SpotBugs%20on%20Jenkins%20%28user%3A%20baloghadamsoftware,%20branch%3A%20PermissionsSuper%29)
[NanoHTTPD](https://codechecker-demo.eastus.cloudapp.azure.com/java-static-analysis/reports?run=SpotBugs%20on%20NanoHTTPD%20%28user%3A%20spotbugs,%20branch%3A%20master%29&newcheck=SpotBugs%20on%20NanoHTTPD%20%28user%3A%20baloghadamsoftware,%20branch%3A%20PermissionsSuper%29)
[Bt](https://codechecker-demo.eastus.cloudapp.azure.com/java-static-analysis/reports?run=SpotBugs%20on%20Bt%20%28user%3A%20spotbugs,%20branch%3A%20master%29&newcheck=SpotBugs%20on%20Bt%20%28user%3A%20baloghadamsoftware,%20branch%3A%20PermissionsSuper%29)
[Ttorrent](https://codechecker-demo.eastus.cloudapp.azure.com/java-static-analysis/reports?run=SpotBugs%20on%20Ttorrent%20%28user%3A%20spotbugs,%20branch%3A%20master%29&newcheck=SpotBugs%20on%20Ttorrent%20%28user%3A%20baloghadamsoftware,%20branch%3A%20PermissionsSuper%29)

**Comment by baloghadamsoftware** [comment]
@KengoTODA or @ThrawnCA do you have any comments on this one? Or should we put it into FindSecBugs instead? What is your opinion, @h3xstream?

File: spotbugs/src/main/java/edu/umd/cs/findbugs/detect/PermissionsSuper.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
+/*
+ * FindBugs - Find bugs in Java programs
```
Comment (ThrawnCA) [inline spotbugs/src/main/java/edu/umd/cs/findbugs/detect/PermissionsSuper.java:2]: Should be SpotBugs not FindBugs.

**Review by ThrawnCA — CHANGES_REQUESTED** [review]

**Review by ThrawnCA — APPROVED** [review]

File: spotbugs/src/main/java/edu/umd/cs/findbugs/detect/PermissionsSuper.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        if (seen == Const.ARETURN) {
+            XMethod origin = stack.getStackItem(0).getReturnValueOf();
+            if (origin != null) {
+                try {
+                    if (getThisClass().getSuperClass().getClassName()
+                            .equals(origin.getClassName()) &&
+                            getMethod().getSignature().equals(origin.getSignature())) {
+                        return;
+                    }
+                } catch (ClassNotFoundException e) {
```
Comment (KengoTODA) [inline spotbugs/src/main/java/edu/umd/cs/findbugs/detect/PermissionsSuper.java:89]: Do we need to call `AnalysisContext.reportMissingClass(e)` here?

File: spotbugs/src/main/java/edu/umd/cs/findbugs/detect/PermissionsSuper.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    public void sawOpcode(int seen) {
+        if (!checkMethod) {
+            return;
+        }
+
+        if (seen == Const.ARETURN) {
+            XMethod origin = stack.getStackItem(0).getReturnValueOf();
+            if (origin != null) {
+                try {
+                    if (getThisClass().getSuperClass().getClassName()
```
Comment (KengoTODA) [inline spotbugs/src/main/java/edu/umd/cs/findbugs/detect/PermissionsSuper.java:84]: The `JavaClass#getSuperClass()` method may return `null`, so we need to check nullness before we use it, or replace `getSuperClass().getClassName()` with `getSuperclassName()`.

https://commons.apache.org/proper/commons-bcel/apidocs/org/apache/bcel/classfile/JavaClass.html#getSuperClass--

**Review by KengoTODA — CHANGES_REQUESTED** [review]
Almost OK, thanks for keeping CHANGELOG updated!

**Comment by baloghadamsoftware** [comment]
@KengoTODA?

**Review by KengoTODA — DISMISSED** [review]
Confirmed that the new implementation can pass `PermissionsSuperSuperNegativeTest` test case. Thanks for your contribution!

**Comment by KengoTODA** [comment]
@baloghadamsoftware please resolve conflict in this PR, then I'll ship the next minor release v1.5.0.

**Comment by baloghadamsoftware** [comment]
> 
> 
> @baloghadamsoftware please resolve conflict in this PR, then I'll ship the next minor release v1.5.0.

Done.

**Review by KengoTODA — APPROVED** [review]

**Review by ThrawnCA — APPROVED** [review]

**Comment by baloghadamsoftware** [comment]
Thanks! Any chance to get it into 4.5.1?

File: spotbugs/src/main/java/edu/umd/cs/findbugs/detect/PermissionsSuper.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+            checkClass = obj.instanceOf(Repository.lookupClass("java.security.SecureClassLoader"));
+        } catch (ClassNotFoundException e) {
+            AnalysisContext.reportMissingClass(e);
+        }
+        super.visit(obj);
+    }
+
+    @Override
+    public void visitAfter(JavaClass obj) {
+        bugAccumulator.reportAccumulatedBugs();
```
Comment (KengoTODA) [inline spotbugs/src/main/java/edu/umd/cs/findbugs/detect/PermissionsSuper.java:58]: I guess we need to reset the value of the `checkClass` field at this timing. Or when a `ClassNotFoundException` is thrown in line 49, the previous value of `checkClass` will be used so the analysis result could be unstable.

**Review by KengoTODA — CHANGES_REQUESTED** [review]

File: spotbugs/src/main/java/edu/umd/cs/findbugs/detect/PermissionsSuper.java (hunk 4 above)
Comment (baloghadamsoftware) [reply spotbugs/src/main/java/edu/umd/cs/findbugs/detect/PermissionsSuper.java:59]: Good catch! Thank you for noticing it!

**Review by ThrawnCA — APPROVED** [review]

File: spotbugs/src/main/java/edu/umd/cs/findbugs/detect/PermissionsSuper.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+    @Override
+    public void sawOpcode(int seen) {
+        if (seen == Const.ARETURN) {
+            XMethod origin = stack.getStackItem(0).getReturnValueOf();
+            if (origin != null) {
+                try {
+                    if (Arrays.stream(getThisClass().getSuperClasses())
+                            .filter(cls -> cls.getClassName().equals(origin.getClassName()))
+                            .findAny() != null &&
```
Comment (KengoTODA) [inline spotbugs/src/main/java/edu/umd/cs/findbugs/detect/PermissionsSuper.java:85]:
```suggestion
                            .anyMatch(cls -> cls.getClassName().equals(origin.getClassName())) &&
```

**Review by KengoTODA — CHANGES_REQUESTED** [review]

**Comment by ThrawnCA** [comment]
This seems to be incorporating a lot of changes from another merge, unrelated to this issue.

**Comment by baloghadamsoftware** [comment]
> This seems to be incorporating a lot of changes from another merge, unrelated to this issue.

Seems to be a GitHub error. If I execute `git diff master` then I only get the changes I have made.

**Comment by baloghadamsoftware** [comment]
> This seems to be incorporating a lot of changes from another merge, unrelated to this issue.

Sorry, now I had to force-push to restore the history. I do not know what went wrong but now it seems to be correct.

**Review by ThrawnCA — APPROVED** [review]

**Review by KengoTODA — APPROVED** [review]
LGTM. A conflict in `CHANGELOG.md` needs to be resolved, but I will handle it in another PR.

**Pull request merged**